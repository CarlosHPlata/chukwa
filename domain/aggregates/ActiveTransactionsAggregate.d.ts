import { Transaction } from "../entities/Transaction";

export type PresentTransactionAggregate = Transaction & {
  total: number;
}

export interface ActiveTransactionsAggregate {
  transactions: PresentTransactionAggregate[];
  currentTotal: number;
  startingTotal: number;
  activeMonthId: number;
}

