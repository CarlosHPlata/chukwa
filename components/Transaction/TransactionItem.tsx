import { PresentTransactionAggregate } from "@/domain/aggregates/ActiveTransactionsAggregate";
import { Transaction } from "@/domain/entities/Transaction";
import useFormattedCurrency from "@/hooks/useFormattedCurrency";
import useFormattedDate from "@/hooks/useFormattedDate";
import MaterialIcons from "@expo/vector-icons/MaterialIcons";
import { StyleSheet, Text, TouchableHighlight, View } from "react-native";

type Props = {
  transaction: PresentTransactionAggregate;
  onPress: (transaction: Transaction) => void;
};

export default function TransactionItem({ transaction, onPress }: Props) {
  const formattedAmount = useFormattedCurrency(transaction.amount);
  const formattedTotal = useFormattedCurrency(transaction.total);
  const formattedDate = useFormattedDate(transaction.date, 'datetime');

  return (
    <TouchableHighlight
      testID="transaction-item:pressable"
      key={transaction.id}
      onPress={() => onPress(transaction)}
      style={styles.transactionContainer}
    >
      <View>
        <View style={{ flexDirection: "row", alignItems: "center" }}>
          <Text style={{ fontSize: 16, flex: 1, fontWeight: "bold" }}>
            {transaction.description}
          </Text>
          <MaterialIcons name={transaction.concept.icon} size={16} />
          <MaterialIcons name={transaction.origin.icon} size={16} />
        </View>
        <View style={{ flexDirection: "row", alignItems: "center"}}>
          <Text style={{ color: transaction.isWithdrawal ? "red" : "green", flex: 1 }}>
            {transaction.isWithdrawal ? "-" : "+"}{formattedAmount}
          </Text>
          <Text style={{ color: "gray", fontSize: 12 }}>
            {formattedTotal}
          </Text>
        </View>
        <Text style={{ color: "gray" }}>
          {formattedDate}
        </Text>
      </View>
    </TouchableHighlight>
  );
}

const styles = StyleSheet.create({
  transactionContainer: {
    backgroundColor: "white",
    padding: 10,
    margin: 5,
    borderRadius: 8,
  },
});
