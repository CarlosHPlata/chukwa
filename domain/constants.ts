export const CURRENCY_PRECISION = 1000; // 1 euro = 1000 cents
export const VIEW_DATETIME_FORMAT = "dd-MM-yyyy HH:mm";
export const VIEW_DATE_FORMAT = "dd-MM-yyyy";
export const SAVE_DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSZZ"; // ISO 8601 format for Luxon
