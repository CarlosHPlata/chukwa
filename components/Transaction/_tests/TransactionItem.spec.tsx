import { PresentTransactionAggregate } from "@/domain/aggregates/ActiveTransactionsAggregate";
import { VIEW_DATETIME_FORMAT } from "@/domain/constants";
import { fireEvent, render } from "@testing-library/react-native";
import { DateTime } from "luxon";
import TransactionItem from "../TransactionItem";

jest.mock('@expo/vector-icons/MaterialIcons', () => {
  const React = require('react');
  return {
    __esModule: true,
    default: ({ name, ...props }: { name: string } & unknown) => {
      return React.createElement('MaterialIcons', { ...props, 'data-icon': name });
    },
  };
});

describe("TransactionItem", () => {
  const date = DateTime.fromISO("2023-10-01T12:00:00Z");
  const mockTransaction: PresentTransactionAggregate = {
    id: "1",
    description: "Test Transaction",
    amount: 100,
    total: 150,
    date: date.toJSDate(),
    isWithdrawal: false,
    concept: { icon: "payment", name: "Payment", id: "1" },
    origin: { icon: "payment", name: "balance", id: "1" },
    userId: ["1"],
  };


  it("renders correctly", () => {
    const { toJSON, getByText } = render(
      <TransactionItem transaction={mockTransaction} onPress={() => {}} />
    );

    expect(getByText("Test Transaction")).toBeTruthy();
    expect(getByText("0,15 â‚¬")).toBeTruthy();
    expect(getByText(date.toFormat(VIEW_DATETIME_FORMAT))).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it("calls onPress when pressed", () => {
    const onPressMock = jest.fn();
    const { getByTestId } = render(
      <TransactionItem transaction={mockTransaction} onPress={onPressMock} />
    );

    fireEvent.press(getByTestId("transaction-item:pressable"));
    expect(onPressMock).toHaveBeenCalledWith(mockTransaction);
  });
});
