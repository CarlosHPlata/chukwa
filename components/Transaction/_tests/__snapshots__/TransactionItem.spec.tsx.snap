// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionItem renders correctly 1`] = `
<View
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "white",
      "borderRadius": 8,
      "margin": 5,
      "padding": 10,
    }
  }
  testID="transaction-item:pressable"
>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Test Transaction
      </Text>
      <MaterialIcons
        data-icon="payment"
        size={16}
      />
      <MaterialIcons
        data-icon="payment"
        size={16}
      />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "color": "green",
            "flex": 1,
          }
        }
      >
        +
        0,10 €
      </Text>
      <Text
        style={
          {
            "color": "gray",
            "fontSize": 12,
          }
        }
      >
        0,15 €
      </Text>
    </View>
    <Text
      style={
        {
          "color": "gray",
        }
      }
    >
      01-10-2023 14:00
    </Text>
  </View>
</View>
`;
